File: packages/store/src/backends/memory.ts
// ─── InMemoryStore ────────────────────────────────────────────────────────
// Reference implementation. Map-based. Synchronous. Zero dependencies.
// For tests and lightweight local use.

import type {
  SyncStore,
  SyncStoreInQuery,
  StoreMeta,
  StoreFilter,
} from "../types.js";
import type {
  AtomicMutationRequest,
  AtomicMutationResult,
  AtomicCompletedMutationResult,
  StoreCondition,
  StoreTarget,
  StoreVersion,
  SyncAtomicMutationStore,
  VersionedStoreValue,
} from "../atomic.js";
import { cloneJson, targetKey, validateAtomicRequest } from "../atomic.js";
import { compareCodePoints } from "../order.js";
import { NON_SCALAR_FILTER_MESSAGE } from "../sql.js";

/** The message the SQLite adapter already raises for a non-scalar `listWhereIn`
 *  value. The reference raises the same one so the two backends agree. */
const NON_SCALAR_IN_MESSAGE = "Store IN queries only support JSON scalar values.";

function isJsonScalar(value: unknown): boolean {
  return (
    value === null ||
    typeof value === "string" ||
    typeof value === "number" ||
    typeof value === "boolean" ||
    typeof value === "bigint"
  );
}

/** There is no deep equality anywhere in the port, so a non-scalar `where` value
 *  is a caller error. The SQLite adapter cannot bind one at all; the reference
 *  used to return `[]` and silently mean "no match". */
function assertScalarWhere(where: Record<string, unknown>): void {
  for (const value of Object.values(where)) {
    if (!isJsonScalar(value)) throw new Error(NON_SCALAR_FILTER_MESSAGE);
  }
}

function assertScalarInValues(values: readonly unknown[]): void {
  for (const value of values) {
    if (!isJsonScalar(value)) throw new Error(NON_SCALAR_IN_MESSAGE);
  }
}

let nextMemoryStoreId = 1;

export class InMemoryStore
  implements SyncStore, SyncStoreInQuery, SyncAtomicMutationStore
{
  readonly meta: StoreMeta = {
    backend: "memory",
  };

  /** Key-value storage. */
  private kv = new Map<string, unknown>();

  /** Collection storage: collection name -> Map of id -> item. */
  private collections = new Map<string, Map<string, unknown>>();

  /** Version metadata is separate from values so deletes never revive tokens. */
  private versions = new Map<string, StoreVersion>();
  private nextVersionNumber = 1;
  private readonly versionPrefix = `m${nextMemoryStoreId++}`;
  private receipts = new Map<
    string,
    {
      requestDigest: string;
      result: AtomicCompletedMutationResult & { status: "applied" };
    }
  >();

  // ── Key-Value ──────────────────────────────────────────────────────

  get<T>(key: string): T | null {
    const value = this.kv.get(key);
    return value !== undefined ? (value as T) : null;
  }

  set<T>(key: string, value: T): void {
    this.kv.set(key, value);
    this.versions.set(targetKey({ kind: "key", key }), this.newVersion());
  }

  has(key: string): boolean {
    return this.kv.has(key);
  }

  delete(key: string): boolean {
    const deleted = this.kv.delete(key);
    if (deleted) this.versions.delete(targetKey({ kind: "key", key }));
    return deleted;
  }

  keys(prefix?: string): string[] {
    const allKeys = [...this.kv.keys()];
    const selected = prefix
      ? allKeys.filter((k) => k.startsWith(prefix))
      : allKeys;
    // Code point ascending, matching the SQLite adapter's `ORDER BY key` under
    // the BINARY collation. Map iteration order (insertion) is not the contract.
    return selected.sort(compareCodePoints);
  }

  // ── Collections ────────────────────────────────────────────────────

  private ensureCollection(name: string): Map<string, unknown> {
    // The SQLite adapter cannot name a table for the empty string and rejects it;
    // the reference used to accept it and quietly serve a collection nobody could
    // address on a persistent backend.
    if (name.length === 0) throw new Error("Invalid collection name");
    let col = this.collections.get(name);
    if (!col) {
      col = new Map<string, unknown>();
      this.collections.set(name, col);
    }
    return col;
  }

  list<T>(collection: string, filter?: StoreFilter): T[] {
    const col = this.ensureCollection(collection);
    let items = [...col.values()] as T[];
    items = applyFilter(items, filter);
    return items;
  }

  listWhereIn<T>(
    collection: string,
    field: string,
    values: readonly unknown[],
    filter?: StoreFilter
  ): T[] {
    if (values.length === 0) return [];
    // Validation order mirrors the SQLite adapter, which builds the caller's
    // WHERE clause before the IN clause.
    if (filter?.where) assertScalarWhere(filter.where);
    assertScalarInValues(values);
    const set = new Set(values);
    const col = this.ensureCollection(collection);
    const items = [...col.values()].filter((item) => {
      if (typeof item !== "object" || item === null) return false;
      return set.has((item as Record<string, unknown>)[field]);
    }) as T[];
    return applyFilter(items, filter);
  }

  getById<T>(collection: string, id: string): T | null {
    const col = this.ensureCollection(collection);
    const item = col.get(id);
    return item !== undefined ? (item as T) : null;
  }

  put<T extends { id: string }>(collection: string, item: T): T {
    const col = this.ensureCollection(collection);
    col.set(item.id, item);
    this.versions.set(
      targetKey({ kind: "record", collection, id: item.id }),
      this.newVersion()
    );
    return item;
  }

  remove(collection: string, id: string): boolean {
    const col = this.ensureCollection(collection);
    const removed = col.delete(id);
    if (removed)
      this.versions.delete(targetKey({ kind: "record", collection, id }));
    return removed;
  }

  count(collection: string, filter?: StoreFilter): number {
    const col = this.ensureCollection(collection);
    if (!filter?.where) return col.size;
    // `count` answers "how many match", so it reads only `where`. Routing through
    // applyFilter would also slice, making `count(c, {where, limit: 1})` return 1
    // while the SQLite adapter (which builds a WHERE clause and nothing else)
    // returns the true total.
    const where = filter.where;
    assertScalarWhere(where);
    let total = 0;
    for (const item of col.values()) {
      if (matchesWhere(item, where)) total++;
    }
    return total;
  }

  getVersioned<T>(target: StoreTarget): VersionedStoreValue<T> | null {
    const value =
      target.kind === "key"
        ? (this.kv.get(target.key) as T | undefined)
        : (this.ensureCollection(target.collection).get(target.id) as
            | T
            | undefined);
    if (value === undefined && !this.hasTarget(target)) return null;
    const version = this.versions.get(targetKey(target));
    // Values written by old code before version metadata existed are assigned a
    // token lazily. This keeps the capability useful for an already-populated
    // in-memory instance while preserving write-order semantics thereafter.
    if (!version) {
      const created = this.newVersion();
      this.versions.set(targetKey(target), created);
      return { value: value as T, version: created };
    }
    return { value: value as T, version };
  }

  mutateAtomically(request: AtomicMutationRequest): AtomicMutationResult {
    const validated = validateAtomicRequest(request);
    const idempotencyKey = validated.idempotency?.key;
    if (idempotencyKey !== undefined) {
      const prior = this.receipts.get(idempotencyKey);
      if (prior) {
        if (prior.requestDigest !== validated.requestDigest) {
          return {
            status: "idempotency-conflict",
            key: idempotencyKey,
            expectedRequestDigest: prior.requestDigest,
            receivedRequestDigest: validated.requestDigest,
          };
        }
        return {
          status: "replayed",
          requestDigest: prior.requestDigest,
          versions: prior.result.versions.map((entry) => ({
            target: cloneTarget(entry.target),
            version: entry.version,
          })),
          ...(prior.result.outcome === undefined
            ? {}
            : { outcome: cloneJson(prior.result.outcome) }),
        };
      }
    }

    // All validation is complete before this point. The reference implementation
    // is synchronous, so no caller can observe an intermediate state while the
    // condition check and operation application run.
    for (const condition of validated.conditions) {
      const observed = this.observe(condition.target);
      if (!conditionMatches(condition, observed)) {
        return {
          status: "conflict",
          condition: cloneCondition(condition),
          observed:
            observed === null
              ? "missing"
              : condition.expected === "version"
              ? observed.version
              : "present",
        };
      }
    }

    const versions: { target: StoreTarget; version: StoreVersion | null }[] =
      [];
    for (const operation of validated.operations) {
      const target = operationTarget(operation);
      switch (operation.op) {
        case "set":
          this.set(operation.key, cloneJson(operation.value));
          versions.push({
            target,
            version: this.versions.get(targetKey(target))!,
          });
          break;
        case "delete":
          this.delete(operation.key);
          versions.push({ target, version: null });
          break;
        case "put":
          this.put(operation.collection, cloneJson(operation.item));
          versions.push({
            target,
            version: this.versions.get(targetKey(target))!,
          });
          break;
        case "remove":
          this.remove(operation.collection, operation.id);
          versions.push({ target, version: null });
          break;
      }
    }
    const applied: AtomicCompletedMutationResult & { status: "applied" } = {
      status: "applied",
      requestDigest: validated.requestDigest,
      versions,
      ...(validated.outcome === undefined
        ? {}
        : { outcome: cloneJson(validated.outcome) }),
    };
    if (idempotencyKey !== undefined) {
      this.receipts.set(idempotencyKey, {
        requestDigest: validated.requestDigest,
        result: {
          ...applied,
          versions: applied.versions.map((entry) => ({
            target: cloneTarget(entry.target),
            version: entry.version,
          })),
          ...(applied.outcome === undefined
            ? {}
            : { outcome: cloneJson(applied.outcome) }),
        },
      });
    }
    return applied;
  }

  private newVersion(): StoreVersion {
    return `${this.versionPrefix}-v${this.nextVersionNumber++}` as StoreVersion;
  }

  private observe(target: StoreTarget): VersionedStoreValue<unknown> | null {
    const value =
      target.kind === "key"
        ? this.kv.get(target.key)
        : this.ensureCollection(target.collection).get(target.id);
    if (value === undefined && !this.hasTarget(target)) return null;
    const version = this.versions.get(targetKey(target));
    if (!version) {
      const created = this.newVersion();
      this.versions.set(targetKey(target), created);
      return { value, version: created };
    }
    return { value, version };
  }

  private hasTarget(target: StoreTarget): boolean {
    return target.kind === "key"
      ? this.kv.has(target.key)
      : this.ensureCollection(target.collection).has(target.id);
  }
}

function cloneTarget(target: StoreTarget): StoreTarget {
  return target.kind === "key"
    ? { kind: "key", key: target.key }
    : { kind: "record", collection: target.collection, id: target.id };
}

function cloneCondition(condition: StoreCondition): StoreCondition {
  return condition.expected === "version"
    ? {
        target: cloneTarget(condition.target),
        expected: "version",
        version: condition.version,
      }
    : { target: cloneTarget(condition.target), expected: condition.expected };
}

function operationTarget(operation: {
  op: string;
  key?: string;
  collection?: string;
  id?: string;
  item?: { id: string };
}): StoreTarget {
  if (operation.op === "set" || operation.op === "delete")
    return { kind: "key", key: operation.key! };
  if (operation.op === "put")
    return {
      kind: "record",
      collection: operation.collection!,
      id: operation.item!.id,
    };
  return {
    kind: "record",
    collection: operation.collection!,
    id: operation.id!,
  };
}

function conditionMatches(
  condition: StoreCondition,
  observed: VersionedStoreValue<unknown> | null
): boolean {
  if (condition.expected === "missing") return observed === null;
  if (condition.expected === "present") return observed !== null;
  return observed !== null && observed.version === condition.version;
}

// ── Filter Logic ───────────────────────────────────────────────────────────

function matchesWhere(item: unknown, where: Record<string, unknown>): boolean {
  if (typeof item !== "object" || item === null) return false;
  const record = item as Record<string, unknown>;
  for (const [key, value] of Object.entries(where)) {
    if (record[key] !== value) return false;
  }
  return true;
}

function applyFilter<T>(items: T[], filter?: StoreFilter): T[] {
  if (!filter) return items;

  let result = items;

  // Where clause — exact match
  if (filter.where) {
    const where = filter.where;
    assertScalarWhere(where);
    result = result.filter((item) => matchesWhere(item, where));
  }

  // Sort — stable, so ties keep insertion order
  if (filter.sortBy) {
    const field = filter.sortBy;
    const dir = filter.sortDir === "desc" ? -1 : 1;
    result = [...result].sort((a, b) => {
      const aVal = (a as Record<string, unknown>)[field];
      const bVal = (b as Record<string, unknown>)[field];
      const aMissing = aVal === undefined || aVal === null;
      const bMissing = bVal === undefined || bVal === null;
      // Null and missing sort LAST in both directions, so the direction
      // multiplier deliberately does not apply to them.
      if (aMissing && bMissing) return 0;
      if (aMissing) return 1;
      if (bMissing) return -1;
      if (aVal === bVal) return 0;
      if (typeof aVal === "string" && typeof bVal === "string") {
        // JS `<` on strings compares UTF-16 code units; SQLite's BINARY
        // collation compares UTF-8 bytes. They differ above the BMP.
        return compareCodePoints(aVal, bVal) * dir;
      }
      return aVal < bVal ? -1 * dir : 1 * dir;
    });
  }

  // Offset, then limit — both after the sort
  if (filter.offset !== undefined && filter.offset > 0) {
    result = result.slice(Math.floor(filter.offset));
  }

  // A negative limit clamps to zero (no rows), matching the SQLite adapter's
  // `LIMIT max(0, floor(limit))`. Returning everything for `limit: -1` was the
  // reference's own reading of "no limit".
  if (filter.limit !== undefined) {
    result = result.slice(0, Math.max(0, Math.floor(filter.limit)));
  }

  return result;
}

File: python/store/src/mirk/store/memory.py
"""In-memory reference implementation of the Mirk store port."""

from __future__ import annotations

import json
from typing import Any, cast

from .atomic import (
    IN_PROCESS_ATOMIC_LIMITS,
    AtomicMutationLimits,
    clone_condition,
    clone_json,
    clone_target,
    condition_matches,
    operation_target,
    resolve_atomic_limits,
    target_key,
    validate_atomic_request,
)
from .filter import IN_SCALAR_MESSAGE, apply_filter, dumps_json, is_scalar, json_equal
from .types import JsonObject, StoreFilter, StoreMeta

__all__ = ["InMemoryStore", "copy_json"]

_MISSING = object()

_next_memory_store_id = 1


def copy_json(value: Any) -> Any:
    """Deep copy through a JSON round trip. Rejects NaN and infinities.

    Encoding through `dumps_json` keeps both backends returning the same
    number type: an integral float comes back as an int here exactly as it
    does from a SQLite row a TypeScript writer produced.
    """
    return json.loads(dumps_json(value))


def _assert_collection(collection: str) -> None:
    if len(collection) == 0:
        raise ValueError("Invalid collection name")


class InMemoryStore:
    """Reference store. Insertion-ordered, copy on write and on read."""

    def __init__(
        self,
        *,
        version_identity: str | None = None,
        atomic_limits: AtomicMutationLimits | dict[str, Any] | None = None,
    ) -> None:
        global _next_memory_store_id
        self._meta = StoreMeta(backend="memory")
        self._kv: dict[str, Any] = {}
        self._collections: dict[str, dict[str, Any]] = {}
        # Version metadata is separate from values so deletes never revive tokens.
        self._versions: dict[str, str] = {}
        self._next_version_number = 1
        if version_identity is None:
            version_identity = f"m{_next_memory_store_id}"
            _next_memory_store_id += 1
        self._version_prefix = version_identity
        self._receipts: dict[str, dict[str, Any]] = {}
        self._atomic_limits = resolve_atomic_limits(atomic_limits, IN_PROCESS_ATOMIC_LIMITS)

    @property
    def meta(self) -> StoreMeta:
        return self._meta

    @property
    def atomic_limits(self) -> AtomicMutationLimits:
        """The request bounds this store enforces."""
        return self._atomic_limits

    # ── Key-value ────────────────────────────────────────────────────────
    def get(self, key: str) -> Any:
        if key not in self._kv:
            return None
        return copy_json(self._kv[key])

    def set(self, key: str, value: Any) -> None:
        self._kv[key] = copy_json(value)
        self._versions[target_key({"kind": "key", "key": key})] = self._new_version()

    def has(self, key: str) -> bool:
        return key in self._kv

    def delete(self, key: str) -> bool:
        if key in self._kv:
            del self._kv[key]
            self._versions.pop(target_key({"kind": "key", "key": key}), None)
            return True
        return False

    def keys(self, prefix: str | None = None) -> list[str]:
        ordered = sorted(self._kv.keys())
        if not prefix:
            return ordered
        return [key for key in ordered if key.startswith(prefix)]

    # ── Collections ──────────────────────────────────────────────────────
    def _collection(self, collection: str) -> dict[str, Any]:
        _assert_collection(collection)
        existing = self._collections.get(collection)
        if existing is None:
            existing = {}
            self._collections[collection] = existing
        return existing

    def list(self, collection: str, filter: StoreFilter | None = None) -> list[Any]:
        items = list(self._collection(collection).values())
        return [copy_json(item) for item in apply_filter(items, filter)]

    def getById(self, collection: str, id: str) -> Any:
        item = self._collection(collection).get(id, _MISSING)
        if item is _MISSING:
            return None
        return copy_json(item)

    def put(self, collection: str, item: JsonObject) -> JsonObject:
        col = self._collection(collection)
        item_id = item.get("id")
        if not isinstance(item_id, str):
            raise ValueError("Collection items must have a string id")
        col[item_id] = copy_json(item)
        self._versions[target_key({"kind": "record", "collection": collection, "id": item_id})] = (
            self._new_version()
        )
        return item

    def remove(self, collection: str, id: str) -> bool:
        col = self._collection(collection)
        if id in col:
            del col[id]
            self._versions.pop(
                target_key({"kind": "record", "collection": collection, "id": id}), None
            )
            return True
        return False

    def count(self, collection: str, filter: StoreFilter | None = None) -> int:
        items = list(self._collection(collection).values())
        where = filter.get("where") if filter else None
        narrowed: StoreFilter = {"where": where} if where else {}
        return len(apply_filter(items, narrowed))

    # ── Atomic mutation ──────────────────────────────────────────────────
    def _new_version(self) -> str:
        version = f"{self._version_prefix}-v{self._next_version_number}"
        self._next_version_number += 1
        return version

    def _has_target(self, target: dict[str, Any]) -> bool:
        if target["kind"] == "key":
            return target["key"] in self._kv
        return target["id"] in self._collection(target["collection"])

    def _observe(self, target: dict[str, Any]) -> dict[str, Any] | None:
        """The stored value and its token, minting one for a value that has none."""
        if target["kind"] == "key":
            value = self._kv.get(target["key"], _MISSING)
        else:
            value = self._collection(target["collection"]).get(target["id"], _MISSING)
        if value is _MISSING and not self._has_target(target):
            return None
        key = target_key(target)
        version = self._versions.get(key)
        # Values written before version metadata existed are assigned a token
        # lazily, which keeps the capability useful for an already-populated
        # instance while preserving write-order semantics thereafter.
        if version is None:
            version = self._new_version()
            self._versions[key] = version
        return {"value": None if value is _MISSING else value, "version": version}

    def getVersioned(self, target: dict[str, Any]) -> dict[str, Any] | None:
        """The value at a target and the token identifying this exact revision."""
        observed = self._observe(target)
        if observed is None:
            return None
        return {"value": copy_json(observed["value"]), "version": observed["version"]}

    def mutateAtomically(self, request: dict[str, Any]) -> dict[str, Any]:
        """Apply conditions and operations as one indivisible step."""
        validated = validate_atomic_request(request, self._atomic_limits)
        key = validated.idempotency_key
        if key is not None:
            prior = self._receipts.get(key)
            if prior is not None:
                if prior["requestDigest"] != validated.request_digest:
                    return {
                        "status": "idempotency-conflict",
                        "key": key,
                        "expectedRequestDigest": prior["requestDigest"],
                        "receivedRequestDigest": validated.request_digest,
                    }
                replayed: dict[str, Any] = {
                    "status": "replayed",
                    "requestDigest": prior["requestDigest"],
                    "versions": [
                        {"target": clone_target(entry["target"]), "version": entry["version"]}
                        for entry in prior["result"]["versions"]
                    ],
                }
                if "outcome" in prior["result"]:
                    replayed["outcome"] = clone_json(prior["result"]["outcome"])
                return replayed

        # All validation is complete before this point. Nothing here yields, so
        # no caller can observe an intermediate state between the condition
        # check and the last operation.
        for condition in validated.conditions:
            observed = self._observe(condition["target"])
            if not condition_matches(condition, observed):
                return {
                    "status": "conflict",
                    "condition": clone_condition(condition),
                    "observed": (
                        "missing"
                        if observed is None
                        else observed["version"]
                        if condition["expected"] == "version"
                        else "present"
                    ),
                }

        versions: list[dict[str, Any]] = []
        for operation in validated.operations:
            target = operation_target(operation)
            op = operation["op"]
            if op == "set":
                self.set(operation["key"], clone_json(operation["value"]))
                versions.append({"target": target, "version": self._versions[target_key(target)]})
            elif op == "delete":
                self.delete(operation["key"])
                versions.append({"target": target, "version": None})
            elif op == "put":
                self.put(operation["collection"], clone_json(operation["item"]))
                versions.append({"target": target, "version": self._versions[target_key(target)]})
            else:
                self.remove(operation["collection"], operation["id"])
                versions.append({"target": target, "version": None})

        applied: dict[str, Any] = {
            "status": "applied",
            "requestDigest": validated.request_digest,
            "versions": versions,
        }
        if validated.has_outcome:
            applied["outcome"] = clone_json(validated.outcome)
        if key is not None:
            receipt: dict[str, Any] = {
                "status": "applied",
                "requestDigest": applied["requestDigest"],
                "versions": [
                    {"target": clone_target(entry["target"]), "version": entry["version"]}
                    for entry in versions
                ],
            }
            if "outcome" in applied:
                receipt["outcome"] = clone_json(applied["outcome"])
            self._receipts[key] = {
                "requestDigest": validated.request_digest,
                "result": receipt,
            }
        return applied

    def listWhereIn(
        self,
        collection: str,
        field: str,
        values: list[Any],
        filter: StoreFilter | None = None,
    ) -> list[Any]:
        if len(values) == 0:
            return []
        for value in values:
            if not is_scalar(value):
                raise ValueError(IN_SCALAR_MESSAGE)
        items = list(self._collection(collection).values())
        matched = [item for item in items if _field_in(item, field, values)]
        return [copy_json(item) for item in apply_filter(matched, filter)]


def _field_in(item: Any, field: str, values: list[Any]) -> bool:
    if not isinstance(item, dict):
        return False
    record = cast(dict[str, Any], item)
    if field not in record:
        return False
    actual = record[field]
    return any(json_equal(actual, value) for value in values)
